File: ts/src/test/Exchange/test.features.ts
import assert from 'assert';
import { Exchange } from "../../../ccxt";
import testSharedMethods from './base/test.sharedMethods.js';

async function testFeatures (exchange: Exchange, skippedProperties: object) {
    const marketTypes = [ 'spot', 'swap', 'future', 'option' ];
    const subTypes = [ 'linear', 'inverse' ];
    const features = exchange.features;
    const keys = Object.keys (features);
    for (let i = 0; i < keys.length; i++) {
        testSharedMethods.assertInArray (exchange, skippedProperties, 'features', keys, i, marketTypes);
        const marketType = keys[i];
        const value = features[marketType];
        // assert (value !== undefined, 'exchange.features["' + marketType + '"] is undefined, that key should be either absent or have a value');
        if (value === undefined) {
            continue;
        }
        if (marketType === 'spot') {
            testFeaturesInner (exchange, skippedProperties, value);
        } else {
            const subKeys = Object.keys (value);
            for (let j = 0; j < subKeys.length; j++) {
                const subKey = subKeys[j];
                testSharedMethods.assertInArray (exchange, skippedProperties, 'features', subKeys, j, subTypes);
                const subValue = value[subKey];
                // sometimes it might not be available for exchange, eg. future>inverse)
                if (subValue !== undefined) {
                    testFeaturesInner (exchange, skippedProperties, subValue);
                }
            }
        }
    }
}

function testFeaturesInner (exchange: Exchange, skippedProperties: object, featureObj: any) {
    const format = {
        'sandbox': false,
        'createOrder': {
            'marginMode': false,
            'triggerPrice': false,
            'triggerPriceType': {
                'mark': false,
                'last': false,
                'index': false,
            },
            'stopLossPrice': false,
            'takeProfitPrice': false,
            'attachedStopLossTakeProfit': {
                'triggerPriceType': {
                    'last': false,
                    'mark': false,
                    'index': false,
                },
                'price': false,
            },
            'timeInForce': {
                'GTC': false,
                'IOC': false,
                'FOK': false,
                'PO': false,
                'GTD': false,
                // 'GTX': false,
            },
            'hedged': false,
            // exchange-supported features
            'trailing': false,
            // 'iceberg': false,
            // 'twap': false,
            // 'selfTradePrevention': false,
            // 'oco': false,
        },
        'createOrders': {
            'max': 5,
        },
        'fetchMyTrades': {
            'marginMode': false,
            'daysBack': 0,
            'limit': 0,
            'untilDays': 0,
            // 'symbolRequired': false,
        },
        'fetchOrder': {
            'marginMode': false,
            'trigger': false,
            'trailing': false,
            // 'symbolRequired': false,
        },
        'fetchOpenOrders': {
            'marginMode': false,
            'limit': 0,
            'trigger': false,
            'trailing': false,
            // 'symbolRequired': false,
        },
        'fetchOrders': {
            'marginMode': false,
            'limit': 0,
            'daysBack': 0,
            'untilDays': 0,
            'trigger': false,
            'trailing': false,
            // 'symbolRequired': false,
        },
        'fetchClosedOrders': {
            'marginMode': false,
            'limit': 0,
            'daysBack': 0,
            'daysBackCanceled': 0,
            'untilDays': 0,
            'trigger': false,
            'trailing': false,
            // 'symbolRequired': false,
        },
        'fetchOHLCV': {
            'limit': 0,
        },
    };
    const featureKeys = Object.keys (featureObj);
    const allMethods = Object.keys (exchange.has);
    for (let i = 0; i < featureKeys.length; i++) {
        testSharedMethods.assertInArray (exchange, skippedProperties, 'features', featureKeys, i, allMethods);
        testSharedMethods.assertStructure (exchange, skippedProperties, 'features', featureObj, format, undefined, true); // deep structure check
    }
}

export default testFeatures;

File: julia/Ccxt/test/exchange/test.features.jl
using Test
using Ccxt
function testFeatures(exchange, skippedProperties)

    marketTypes = ["spot", "swap", "future", "option"];
    subTypes = ["linear", "inverse"];
    features = get(exchange, Symbol("features"), nothing);
    keys_var = objectKeys(features);
    i = 0
    while functions.ccxtruthy(functions.ccxt_lt(i, length(keys_var)))
        assertInArray(testSharedMethods, exchange, skippedProperties, "features", keys_var, i, marketTypes);
        marketType = get(keys_var, i + 1, nothing);
        value = get(features, Symbol(marketType), nothing);
        if functions.ccxtruthy(value == nothing)
            i += 1; continue
        end
        if functions.ccxtruthy(marketType == "spot")
            testFeaturesInner(exchange, skippedProperties, value);
        else
            subKeys = objectKeys(value);
            j = 0
            while functions.ccxtruthy(functions.ccxt_lt(j, length(subKeys)))
                subKey = get(subKeys, j + 1, nothing);
                assertInArray(testSharedMethods, exchange, skippedProperties, "features", subKeys, j, subTypes);
                subValue = get(value, Symbol(subKey), nothing);
                if functions.ccxtruthy(subValue != nothing)
                    testFeaturesInner(exchange, skippedProperties, subValue);
                end
                j += 1
            end
        end
        i += 1
    end
    return true
end


function testFeaturesInner(exchange, skippedProperties, featureObj)

    format = Dict{Symbol, Any}(
        Symbol("sandbox") => false,
        Symbol("createOrder") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("triggerPrice") => false,
            Symbol("triggerPriceType") => Dict{Symbol, Any}(
                Symbol("mark") => false,
                Symbol("last") => false,
                Symbol("index") => false
            ),
            Symbol("stopLossPrice") => false,
            Symbol("takeProfitPrice") => false,
            Symbol("attachedStopLossTakeProfit") => Dict{Symbol, Any}(
                Symbol("triggerPriceType") => Dict{Symbol, Any}(
                    Symbol("last") => false,
                    Symbol("mark") => false,
                    Symbol("index") => false
                ),
                Symbol("price") => false
            ),
            Symbol("timeInForce") => Dict{Symbol, Any}(
                Symbol("GTC") => false,
                Symbol("IOC") => false,
                Symbol("FOK") => false,
                Symbol("PO") => false,
                Symbol("GTD") => false
            ),
            Symbol("hedged") => false,
            Symbol("trailing") => false
        ),
        Symbol("createOrders") => Dict{Symbol, Any}(
            Symbol("max") => 5
        ),
        Symbol("fetchMyTrades") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("daysBack") => 0,
            Symbol("limit") => 0,
            Symbol("untilDays") => 0,
            Symbol("symbolRequired") => false
        ),
        Symbol("fetchOrder") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("trigger") => false,
            Symbol("trailing") => false,
            Symbol("symbolRequired") => false
        ),
        Symbol("fetchOpenOrders") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("limit") => 0,
            Symbol("trigger") => false,
            Symbol("trailing") => false,
            Symbol("symbolRequired") => false
        ),
        Symbol("fetchOrders") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("limit") => 0,
            Symbol("daysBack") => 0,
            Symbol("untilDays") => 0,
            Symbol("trigger") => false,
            Symbol("trailing") => false,
            Symbol("symbolRequired") => false
        ),
        Symbol("fetchClosedOrders") => Dict{Symbol, Any}(
            Symbol("marginMode") => false,
            Symbol("limit") => 0,
            Symbol("daysBack") => 0,
            Symbol("daysBackCanceled") => 0,
            Symbol("untilDays") => 0,
            Symbol("trigger") => false,
            Symbol("trailing") => false,
            Symbol("symbolRequired") => false
        ),
        Symbol("fetchOHLCV") => Dict{Symbol, Any}(
            Symbol("limit") => 0
        )
    );
    featureKeys = objectKeys(featureObj);
    allMethods = objectKeys(get(exchange, Symbol("has"), nothing));
    i = 0
    while functions.ccxtruthy(functions.ccxt_lt(i, length(featureKeys)))
        assertInArray(testSharedMethods, exchange, skippedProperties, "features", featureKeys, i, allMethods);
        assertStructure(testSharedMethods, exchange, skippedProperties, "features", featureObj, format, nothing, true);
        i += 1
    end
end
